django/db/backends/postgresql/schema.py

```diff
@@ -94,3 +94,41 @@ def _alter_column_type_sql(self, table, old_field, new_field, new_type):
             return super(DatabaseSchemaEditor, self)._alter_column_type_sql(
                 table, old_field, new_field, new_type
             )
+
+    def _alter_field(self, model, old_field, new_field, old_type, new_type,
+                     old_db_params, new_db_params, strict=False):
+        super(DatabaseSchemaEditor, self)._alter_field(
+            model, old_field, new_field, old_type, new_type, old_db_params,
+            new_db_params, strict
+        )
+        # Added an index?
+        if ((not old_field.db_index and new_field.db_index) or (not old_field.unique and new_field.unique)):
+            db_type = new_field.db_type(connection=self.connection)
+            if db_type is not None and (new_field.db_index or new_field.unique):
+                # Fields with database column types of `varchar` and `text` need
+                # a second index that specifies their operator class, which is
+                # needed when performing correct LIKE queries outside the
+                # C locale. See #12234.
+                #
+                # This code resambles the one found in `self._model_indexes_sql`.
+                # The only difference is that here the statements are
+                # executed immediately.
+                #
+                # The same doesn't apply to array fields such as varchar[size]
+                # and text[size], so skip them.
+                if '[' in db_type:
+                    return
+                if db_type.startswith('varchar'):
+                    self.execute(self._create_index_sql(
+                        model, [new_field], suffix='_like', sql=self.sql_create_varchar_index))
+                elif db_type.startswith('text'):
+                    self.execute(self._create_index_sql(
+                        model, [new_field], suffix='_like', sql=self.sql_create_text_index))
+        # Removed an index?
+        if ((not new_field.db_index and old_field.db_index) or (not new_field.unique and old_field.unique)):
+            # Find the index for this field.
+            # This is needed because the BaseDatabaseSchemaEditor._alter_field
+            # method doesn't drop the '_like' indexes.
+            index_names = self._constraint_names(model, [old_field.column], index=True)
+            for index_name in index_names:
+                self.execute(self._delete_constraint_sql(self.sql_delete_index, model, index_name))
```

Comment:
We shouldn't systematically drop all indexes as this will also remove the ones a `RunSQL` operation could have added. We should only remove the one created by Django.